src/flashcart/ed64/ed64_ll.c

```diff
@@ -0,0 +1,409 @@
+    if (ram_bank)save_cfg |= SAV_RAM_BANK;
+    save_cfg |= SAV_RAM_BANK_APPLY;
+
+    ed64_ll_reg_write(REG_SAV_CFG, save_cfg);
+
+}
+
+void ed64_ll_set_sram_bank (uint8_t bank) {
+
+    ed64_ll_sram_bank = bank == 0 ? 0 : 1;
+
+}
+
+
+void pi_initialize (void) {
+
+	dma_wait();
+	io_write(PI_STATUS_REG, 0x03);
+
+}
+
+// Inits PI for sram transfer
+void pi_initialize_sram (void) {
+
+	io_write(PI_BSD_DOM2_LAT_REG, 0x05);
+	io_write(PI_BSD_DOM2_PWD_REG, 0x0C);
+	io_write(PI_BSD_DOM2_PGS_REG, 0x0D);
+	io_write(PI_BSD_DOM2_RLS_REG, 0x02);
+
+}
+
+void pi_dma_from_sram (void *dest, unsigned long offset, unsigned long size) {
+
+	io_write(PI_DRAM_ADDR_REG, K1_TO_PHYS(dest));
+	io_write(PI_CART_ADDR_REG, (0xA8000000 + offset));
+	 asm volatile ("" : : : "memory");
+	io_write(PI_WR_LEN_REG, (size - 1));
+	 asm volatile ("" : : : "memory");
+
+}
+
+
+void pi_dma_to_sram (void *src, unsigned long offset, unsigned long size) {
+
+	dma_wait();
+
+	io_write(PI_STATUS_REG, 2);
+	io_write(PI_DRAM_ADDR_REG, K1_TO_PHYS(src));
+	io_write(PI_CART_ADDR_REG, (0xA8000000 + offset));
+	io_write(PI_RD_LEN_REG, (size - 1));
+
+}
+
+void pi_dma_from_cart (void* dest, void* src, unsigned long size) {
+
+	dma_wait();
+
+	io_write(PI_STATUS_REG, 0x03);
+	io_write(PI_DRAM_ADDR_REG, K1_TO_PHYS(dest));
+	io_write(PI_CART_ADDR_REG, K0_TO_PHYS(src));
+	io_write(PI_WR_LEN_REG, (size - 1));
+
+}
+
+
+void pi_dma_to_cart (void* dest, void* src, unsigned long size) {
+
+	dma_wait();
+
+	io_write(PI_STATUS_REG, 0x02);
+	io_write(PI_DRAM_ADDR_REG, K1_TO_PHYS(src));
+	io_write(PI_CART_ADDR_REG, K0_TO_PHYS(dest));
+	io_write(PI_RD_LEN_REG, (size - 1));
+
+}
+
+
+// Wrapper to support unaligned access to memory
+void pi_dma_from_cart_safe (void *dest, void *src, unsigned long size) {
+
+	if (!dest || !src || !size) return;
+
+	unsigned long unalignedSrc  = ((unsigned long)src)  % 2;
+	unsigned long unalignedDest = ((unsigned long)dest) % 8;
+
+	//FIXME: Do i really need to check if size is 16bit aligned?
+	if (!unalignedDest && !unalignedSrc && !(size % 2)) {
+		pi_dma_from_cart(dest, src, size);
+		dma_wait();
+
+		return;
+	}
+
+	void* newSrc = (void*)(((unsigned long)src) - unalignedSrc);
+	unsigned long newSize = (size + unalignedSrc) + ((size + unalignedSrc) % 2);
+
+	unsigned char *buffer = memalign(8, newSize);
+	pi_dma_from_cart(buffer, newSrc, newSize);
+	dma_wait();
+
+	memcpy(dest, (buffer + unalignedSrc), size);
+
+	free(buffer);
+
+}
+
+
+int ed64_ll_get_sram_128 (uint8_t *buffer, int size) {
+
+    dma_wait();
+
+    io_write(PI_BSD_DOM2_LAT_REG, 0x05);
+    io_write(PI_BSD_DOM2_PWD_REG, 0x0C);
+    io_write(PI_BSD_DOM2_PGS_REG, 0x0D);
+    io_write(PI_BSD_DOM2_RLS_REG, 0x02);
+
+    dma_wait();
+
+    pi_initialize();
+
+    dma_wait();
+
+    // Offset Large RAM size 128KiB with a 16KiB nudge to allocate enough space
+    // We do this because 128Kib is not recognised and a 32KiB allocates too little 
+    pi_dma_from_sram(buffer, -(size - KiB(16)), size);
```

Comment:
What is happening here? Why second argument is negative? I really don't understand what this offset does.